app/src/main/java/com/futurice/freesound/feature/common/ui/adapter/RecyclerViewAdapter.java

```diff
@@ -0,0 +1,104 @@
+package com.futurice.freesound.feature.common.ui.adapter;
+
+import com.futurice.freesound.common.utils.AndroidPreconditions;
+import com.futurice.freesound.feature.common.DisplayableItem;
+import com.futurice.freesound.inject.fragment.FragmentScope;
+
+import android.support.annotation.NonNull;
+import android.support.v7.util.DiffUtil;
+import android.support.v7.widget.RecyclerView;
+import android.view.ViewGroup;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.Executor;
+import java.util.concurrent.Executors;
+
+import javax.inject.Inject;
+
+import io.reactivex.Observable;
+import io.reactivex.android.schedulers.AndroidSchedulers;
+import io.reactivex.schedulers.Schedulers;
+
+/**
+ * Implementation of {@link android.support.v7.widget.RecyclerView.Adapter} for {@link
+ * DisplayableItem}.
+ */
+@FragmentScope
+public final class RecyclerViewAdapter extends RecyclerView.Adapter {
+
+    @NonNull
+    private final List<DisplayableItem> mItems = new ArrayList<>();
+
+    @NonNull
+    private final IListItemComparator mComparator;
+
+    @NonNull
+    private final IViewHolderFactory mInstantiator;
+
+    @NonNull
+    private final IViewHolderBinder<DisplayableItem> mBinder;
+
+    private Executor mExecutor = Executors.newSingleThreadExecutor();
+
+    @Inject
+    RecyclerViewAdapter(@NonNull final IListItemComparator comparator,
+                        @NonNull final IViewHolderFactory instantiator,
+                        @NonNull final IViewHolderBinder<DisplayableItem> binder) {
+        mComparator = comparator;
+        mInstantiator = instantiator;
+        mBinder = binder;
+    }
+
+    @Override
+    public RecyclerView.ViewHolder onCreateViewHolder(final ViewGroup parent, final int viewType) {
```

Comment:
For public API it's worth declaring and enforcing `@NonNull` for the parameters - this will make it more straightforward if/when we extract it to separate module.